src/validation.cpp

```diff
@@ -3622,17 +3622,41 @@ bool BlockManager::LoadBlockIndex(
         vSortedByHeight.push_back(std::make_pair(pindex->nHeight, pindex));
     }
     sort(vSortedByHeight.begin(), vSortedByHeight.end());
+
+    // Find start of assumed-valid region.
+    int first_assumed_valid_height = std::numeric_limits<int>::max();
+
+    for (const auto& [height, block] : vSortedByHeight) {
+        if (block->IsAssumedValid()) {
+            auto chainstates = chainman.GetAll();
+
+            // assert that the assumed valid blocks come from some assumed-valid chainstate.
+            assert(any_of(chainstates.cbegin(), chainstates.cend(),
+                [](auto chainstate) { return chainstate->reliesOnAssumedValid(); }));
```

Comment:
I don't understand your comment: these two asserts are only performed once if there is an assumed-valid block detected, and then the loop is broken out of (line 3642). We don't want to do the asserts if there is no assumed-valid block, so I don't see how or why we'd do them outside of the loop.

Reply:
Unless I have a misunderstanding here, what I meant: We only have blocks that are assumed valid if we have a chainstate that relies on assume valid, those two will always go together. We will probably have very few chain states and we have a lot of blocks. So it seems to me that it would be more efficient to first loop the chainstates and then the blocks.